src/Datadog.Trace/SpanContext.cs

```diff
@@ -52,10 +52,11 @@ internal SpanContext(ulong? traceId, ulong spanId, SamplingPriority? samplingPri
         /// <param name="parent">The parent context.</param>
         /// <param name="traceContext">The trace context.</param>
         /// <param name="serviceName">The service name to propagate to child spans.</param>
-        internal SpanContext(ISpanContext parent, ITraceContext traceContext, string serviceName)
+        /// <param name="spanId">The propagated span id.</param>
+        internal SpanContext(ISpanContext parent, ITraceContext traceContext, string serviceName, ulong? spanId = null)
             : this(parent?.TraceId, serviceName)
         {
-            SpanId = SpanIdGenerator.ThreadInstance.CreateNew();
+            SpanId = spanId ?? SpanIdGenerator.ThreadInstance.CreateNew();
```

Comment:
Under what circumstance would we have a span id already? Is this for testing purposes?

Reply:
`GetRequestStream` and `GetResponse` are different operations that can be called from different points in the user code. Headers can only be set during `GetRequestStream`.

I started by creating a Span in `GetRequestStream` and retrieving it from `GetResponse` but then we got worried about what could happen if:
 - An error occurs in-between and `GetResponse` is never called
 - `GetResponse` is called from a different thread (without the active span)

To address those concerns, I used a different strategy:
- In `GetRequestStream`, I create a `SpanContext` and injects the headers in the request. I do not change the active span
- In `GetResponse` I look for those headers in the request. If they exist, I create the span from those headers (hence the need of manually setting the span id) and set it as active. Otherwise I run the old logic.